internal/storage/chunk_uploader.go

```diff
@@ -0,0 +1,202 @@
+package storage
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+
+	"cloud.google.com/go/storage"
+	"github.com/googlecloudplatform/gcsfuse/internal/logger"
+	"github.com/googlecloudplatform/gcsfuse/internal/storage/gcs"
+	"github.com/googlecloudplatform/gcsfuse/internal/storage/storageutil"
+	"google.golang.org/api/googleapi"
+)
+
+type UploaderState int
+
+const (
+	// Initialized is the state of successful initialization, no writes so far.
+	Initialized UploaderState = iota
+	// Waiting is the state when no write is in progress, but there have been writes in the past.
+	Waiting
+	// Writing is the state when an asynchronous write is in progress.
+	Writing
+	// WriteError is the state when an Upload failed.
+	WriteError
+	// Destroyed is the state of an uploader which has been finalized.
+	Destroyed
+)
+
+// A chunkUploader is an implementation of ChunkUploader
+// interface, which uses storage.Writer from go-storage-client
+// for resumable upload.
+//
+// It also stores the current state of the uploader.
+type chunkUploader struct {
+	// Internal objects for business logic.
+	wc         *storage.Writer
+	objectName string
+	mu         sync.Mutex
+
+	// Attributes for providing updates to user.
+	totalWriteInitiatedSoFar int64
+	totalWriteSucceededSoFar int64
+	userProgressFunc         func(int64)
+
+	// Internal state for lifecycle management.
+	state UploaderState
+}
+
+// NewChunkUploader creates a new instance of chunkUploader,
+// for the given inputs.
+func NewChunkUploader(ctx context.Context, obj *storage.ObjectHandle, req *gcs.CreateObjectRequest, writeChunkSize int, progressFunc func(int64)) (gcs.ChunkUploader, error) {
+	if ctx == nil {
+		return nil, fmt.Errorf("ctx is nil")
+	}
+	if obj == nil || req == nil {
+		return nil, fmt.Errorf("nil obj or req")
+	}
+
+	if obj.ObjectName() != req.Name {
+		return nil, fmt.Errorf("names of passed object-handle and createObjectRequest.Name don't match: object-handle=%s createObjectRequest.Name=%s", obj.ObjectName(), req.Name)
```

Comment:
let's use camelcase for ObjectHandle as well

Reply:
Done.

Reply:
there is one more object-handle

Reply:
Removed.